Make the shortcut provider optional

youknowriad opened the pull request:
Partially reverts #34539
Related #53874

## What and why?

Gutenberg can be used as a platform/framework to build block editors. Mainly thanks to the @wordpress/block-editor package. That said, the experience today is not as straightforward as it can be. There can be a lot of small gotchas and hacks you need to do in order to achieve the desired result. One of these small things is the need to manually render `ShortcutProvider` component, this PR bundles this behavior within the BlockEditorProvider component, that way custom block editors don't have to render this extra component.

The original PR says that it only impacts "focus loss" behavior but that is not true, if you navigate in the browser between different pages in WP-Admin, focus is moved to the "body" when you enter new pages and some keyboard shortcuts are expected to work in that context directly without the need to focus any part of the UI. For instance the global command palette shortcut. This means that the current PR also fixes a bug. 

## Testing Instructions

1- Check that the different keyboard shortcuts still work as intended. I'm honestly not 100% confident about the impact here but I believe it's a good improvement.

youknowriad committed:
Make the shortcut provider optional

jorgefilipecosta commented on packages/edit-site/src/components/app/index.js:
```diff
@@ -35,16 +34,14 @@ export default function App() {
 	}
 
 	return (
-		<ShortcutProvider style={ { height: '100%' } }>
```
It seems like this code was relying on ShortcutProvider having a div. Not sure if we should not add a div, but if things work I guess we can keep this change.

youknowriad replied:
Yeah I think this height was there to ensure the editor was full height and the removal of the height is probably going to have the same effect. Of course checking that the site editor is still full height.

jorgefilipecosta reviewed (commented):
These end to end test failure may be relevant:
-  [chromium] › editor/various/shortcut-focus-toolbar.spec.js:17:2 › Focus toolbar shortcut (alt + F10) › Focuses correct toolbar in default view options in edit mode 
    Error: expect(received).toBeFocused()
- [chromium] › editor/blocks/heading.spec.js:183:2 › Heading › should change heading level with keyboard shortcuts

jorgefilipecosta reviewed (commented):
The change seems good but I'm able to reproduce the end to end test failures with manual testing.
On Mac option + f10 (may need to press fn) on trunk focus the toolbar in this branch it does not. On Mac in trunk control + option + (1,2,3,4,5,6) changes heading in this branch, it does not.

youknowriad commented:
So it's actually the iframe that is preventing the events from propagating to the parent event handler. In trunk, there's the same problem for real DOM events but React does some trickery to bubble these as react events (when using onKeyDown props...). I wonder if we should just find a way to bubble these in the iframe instead of forcing a provider.

youknowriad committed:
Bubble keydown events

youknowriad commented:
It turns out that we were already bubbling some event types through the iframe, I've added the keydown events to the list, let's see if it fixes the tests. At least it fixed the focusToolbar in my tests.

youknowriad committed:
Prevent double region navigation

youknowriad committed:
Prevent bubbling when navigation mode is handled

youknowriad committed:
Wait to register the event handler

jsnajdr commented on packages/keyboard-shortcuts/src/context.js:
```diff
@@ -3,4 +3,13 @@
+			keyboardShortcut( event );
+		}
+	} );
+} );
```
Shouldn't this be happening in some effect, e.g., in `KeyboardShortcuts.Register`?

This code will be crashing in SSR environment, it should be at least wrapped by a `typeof document !== 'undefined'` check.

Also, if this module ever gets loaded dynamically (like when bundled into a code-split webpack chunk), the `DOMContentLoaded` event will have had already happened.

youknowriad replied:
It can't happen in an effect because there's no effect if you don't have any provider. Maybe I can move the code to the first `registry.add` call.

jsnajdr replied:
> there's no effect if you don't have any provider

But that only means that there is a good design reason to have a provider. And it doesn't need to be exactly a provider, the existing `KeyboardShortcuts.Register` component is a perfect fit for this listener-registering effect. It already has an effect that registers stuff in the data store.

youknowriad replied:
`KeyboardShortcuts.Register` is a "user" code not something that is provided by the library `keyboard-shortcuts`.

The only thing the library provides now are "registerShortcut" actions and "useShortcut" hooks. If you mean `useShortcut` should trigger the "global event listener" if it's not already listening, that's basically the same as overriding the "add" function to do that.

jsnajdr replied:
> If you mean `useShortcut` should trigger the "global event listener" if it's not already listening

Yes, that should work. My request is to make the listener controlled by React code. It should be added when a certain React component or hook is mounted, and automatically removed when it's unmounted. A reference-counted global variable and an effect can make it work:
```js
let listenerCounter = 0;

function useShortcut() {
  useEffect( () => {
    if ( listenerCounter++ === 0 ) {
      document.body.addListener();
    }
    return () => {
      if ( --listenerCounter === 0 ) {
        document.body.removeListener();
      }
    };
  }, [] );
  /* rest of useShortcut code */
}
```

youknowriad replied:
@jsnajdr Why a lingering global event listener is a problem for you?

jsnajdr replied:
The global listener currently has two practical issues: the static module-level access to `document` will break in SSR environments, and the registration on `DOMContentLoaded` event makes some assumptions that don't need to be true.

Code that's triggered by merely importing a module, as a side-effect, without ever using any of its exports, has a propensity to cause trouble for the module users. If you disagree, you can keep the global listener, but I think that at least the two issues mentioned should be fixed.

youknowriad replied:
No, I agree with these two downsides, what I'm wondering is why not just hook the event on "add" in the "context" instead of moving this responsibility to a single useShortcut call which is not global right.

jsnajdr replied:
> why not just hook the event on "add" in the "context"

You mean extending the `Set.prototype.add` method that is called in `useShortcut`, inside `useEffect`, as `shortcuts.current.add()`? Yes, that's fine, because it's exactly the same thing as I proposed above, only slightly reshuffled. Or isn't it? To detect the "first `add`" and "last `delete`", you'll also need the reference counter. I think we are in a complete agreement 🙂

youknowriad replied:
Updated

youknowriad committed:
SSR support for keyboard shortcuts

jsnajdr commented on packages/keyboard-shortcuts/src/context.js:
```diff
@@ -3,4 +3,26 @@
+};
+
+export const context = createContext( {
+	current: {
```
I think we can eliminate the `.current` nested object. This is not a ref, it's an object exposed through a context.

jsnajdr commented on packages/keyboard-shortcuts/src/context.js:
```diff
@@ -3,4 +3,26 @@
+			if ( globalShortcuts.size === 0 ) {
+				document.addEventListener( 'keydown', globalListener );
+			}
+			globalShortcuts.add( shortcut );
```
Very nice, more elegant than the code snippet I proposed. I like how the `Set` is also used as a reference counter.

jsnajdr reviewed (commented):
The `useShortcut` function will need to updated, too, because the method on the context object is no longer `delete`, but `remove`.

There are two additional nice-to-have improvements on `useShortcut`:
- set the `callbackRef.current` value in `useEffect` rather than during render. That's the canonical recommended approach to do this typical task.
- the `isDisabled` prop should default to `false`. Then the `isDisabled` value inside the function doesn't alternate between `undefined` and `false`, possibly rerunning the effect even though it doesn't need to.

youknowriad committed:
Small improvements

youknowriad committed:
Small improvements

jsnajdr commented on packages/keyboard-shortcuts/src/hooks/use-shortcut.js:
```diff
@@ -17,11 +17,18 @@ import { context } from '../context';
+
+	useEffect( () => {
+		callbackRef.current = callback;
+	} );
```
Better to add the `[ callback ]` deps array, although it works fine even without it.

jsnajdr commented on packages/keyboard-shortcuts/src/hooks/use-shortcut.js:
```diff
@@ -34,9 +41,9 @@ export default function useShortcut( name, callback, { isDisabled } = {} ) {
+			shortcuts.delete( _callback );
 		};
-	}, [ name, isDisabled ] );
+	}, [ name, isDisabled, shortcuts ] );
```
It good that `shortcuts` ended up being a dependency. If the context value changes, then all shortcuts will be removed from the old registry and added to the new one, automatically.

youknowriad committed:
Small update

youknowriad committed:
Fix e2e tests

youknowriad commented:
I've noticed that a lot e2e test failures were because events were triggering twice now (it started happening when I added the bubbling of keydown events from the iframe to the parent component) and I've fixed most of these by just adding `stopPropagation` call on the event handler.

So I'm wondering:

 - Why are we catching the event twice, it's almost like the `KeyboardShortcutContext` is registered twice even though I'm not using `ShortcutProvider` at all so in theory, we should only have the "global listener". 

 - Maybe the iframe bubbling code has a bug that forwards the event twice?

Any idea @jsnajdr @ellatrix

jsnajdr commented:
I'll try to debug this. I suspect that it has something to do with how React attaches DOM listeners. When there is a React element with a React event listener:
```jsx
<div onKeydown={ handleKeydown } />
```
then the real DOM listener is not attached on the `div`, but on the React root element, and React uses `event.target` to figure out that it should call the `handleKeydown` function.

That affects the order in which handlers are called during the capture and bubbling phases.

jsnajdr commented:
Some findings. The e2e failure occurs when a custom completer is active, it looks like this:

<img width="498" alt="Screenshot 2023-08-30 at 12 46 46" src="https://github.com/WordPress/gutenberg/assets/664258/7c004b62-8793-4524-8ba8-3ae801c639fb">

Now, when I press Escape, the completer popover disappears, and the paragraph element now _loses focus_. Before this PR, the `document.activeElement` was the `p` element all the time, now the focus goes to the `body` element after Escape. Further typing doesn't go into the paragraph block, but the e2e test expects that it goes.

This is caused by the newly introduced bubbling of the `keydown` events. When I disable the `frameElement.dispatchEvent` call, the bug disappears.

And it has nothing to do with the global `keydown` listener for shortcuts. If I disable the listener, the bug is still there.

I'll continue searching for the code that causes the focus loss.

ellatrix commented on packages/block-editor/src/components/block-tools/index.js:
```diff
@@ -71,44 +71,51 @@ export default function BlockTools( {
 			const clientIds = getSelectedBlockClientIds();
 			if ( clientIds.length ) {
 				event.preventDefault();
+				event.stopPropagation();
```
Why are we stopping propagation? This is usually a sign that something is wrong. You're preventing components up the tree from listening to this event.

youknowriad replied:
I'm explaining this here, https://github.com/WordPress/gutenberg/pull/53942#issuecomment-1697774962 I think we have a bug in the event bubbling code that make the event handler trigger twice for some reason unless  the event propagation is stopped. I don't understand why yet.

ellatrix commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -70,7 +70,7 @@ function bubbleEvents( doc ) {
 	}
 
-	const eventTypes = [ 'dragover', 'mousemove' ];
+	const eventTypes = [ 'dragover', 'mousemove', 'keydown' ];
```
We should definitely not add more event names to this list. Ideally we should removing this bubbling entirely.

youknowriad replied:
If we're totally against this, the alternative would be to add a `ShortcutProvider` to the `Iframe` component to catch the keyboard shortcuts events. The problem though is that right now `ShortcutProvider` also serves as a registry for event handlers `useShortcut` calls, so we need to have two kind of providers:

 - One that keeps track of active event handlers.
 - One that only catches event and try to match them with active event handlers.

jsnajdr commented:
> I'll continue searching for the code that causes the focus loss.

So, the offending focus call is on this line:

https://github.com/WordPress/gutenberg/blob/37dbd03cc1e3b7098aa4047e03994ba69478f7ad/packages/block-editor/src/components/block-tools/index.js#L124

It's handler for the "Clear selection" keyboard shortcut that's registered, on the Esc key, here:

https://github.com/WordPress/gutenberg/blob/37dbd03cc1e3b7098aa4047e03994ba69478f7ad/packages/block-editor/src/components/keyboard-shortcuts/index.js#L87-L94

Now when we started forwarding the `keydown` event to the iframe's parent, the event is caught by the `BlockTools` component, which is the `iframe`'s parent.

ellatrix commented:
I don't understand form the code what exactly ShortcutProvider is being replaced with. In any case, we need a div so that the shortcuts are scoped to that div (shortcuts shouldn't be global). So either we need to move this into BlockEditorProvider or BlockTools . BlockTools (it already has a div that has an onKeyDown listener for other shortcuts) could work too because InspectorControls actually contains a slot from within the block, so shortcuts should continue to work in there.

youknowriad commented:
> I don't understand form the code what exactly ShortcutProvider is being replaced with. In any case, we need a div so that the shortcuts are scoped to that div (shortcuts shouldn't be global). 

I disagree with this take, some shortcuts don't need a div, like command palette. Maybe some block editor shortcuts need a div but right now in trunk that div is a random div containing the whole page and not something scoped to the block editor.

youknowriad commented:
@jsnajdr interesting find.

The question is: in "trunk" that event handler "core/block-editor/unselect" is also supposed to be triggered when you hit "Escape" (it's just caught differently using a global React listener and not DOM listener). So I'd say that probably there's some code that "prevents bubbling" or "prevents default" already in place but that code only run when the auto-complete is open. 

So IMO, the problem is still the same as I said above: the event is kind of triggered twice, one of them is canceled/preventDefault/preventPropagation by the autocomplete but then the autocomplete closes and the second event is not cancelled causing the selection clear.

youknowriad commented:
Ok after a discussion with @ellatrix we figured that iframe event bubbling results in "two React event handlers" being triggered because of React's internal behavior and the new event the bubbling causes. 

and `BlockTools` is relying on its own react event handler and does not use the "Shortcuts" package default one (which is a DOM one), so it creates the double events issue.

This kind of forces to rethink this PR a bit. My current plan is the following:


 - Replace iframe event bubbling with a `ShortcutEventProvider` that catches all events and forward them to the registered handlers.
 - Find a way to keep the global event handler on body.
 - Potentially find a way to keep the scoped event handlers (the event handler of BlockTools is scoped), maybe by passing a "ref" as a scope argument to `useShortcut` call.

ellatrix commented:
Thanks @youknowriad!

youknowriad committed:
Try scoped useShortcut

youknowriad commented:
Pushed a commit to implement the proposed solution (kept the iframe bubbling temporarily). I'm not entirely satisfied for a few reasons:

 - I don't like passing "ref objects" as arguments to hooks because these are dependent on component lifecycle for instance if the element unmounts, the ref is no longer valid.
 - I noticed that several of our block editor shortcuts (removal, inserter after, insert before, duplicate) are actually duplicated in the code base between two or more places. I suspect that it's because we have scoping in place, these shortcuts would have worked if they were not scope. Maybe there's another way to handle these shortcuts or share their implementation. The existing `BlockActions` component was meant to share this kind of logic but it's failing entirely. (probably out of scope of this PR though).
 - I like that this PR basically allows us to "remove" the `__unstableUseShortcutEventMatch` API.

youknowriad commented:
@ellatrix After thinking a bit here. I actually didn't feel great with the new "scope" API especially because of the weird refs passed around that it forces us to do. (Explained a bit better above)

So I decided to give the initial approach of this PR a second chance: iframe event bubbling but keeping the scoping you had in place in BlockTools and similar using "local" event handlers.

I managed to actually fix the problem we had with iframe event bubbling creating two react synthetic events, it's actually as simple as "prevent propagation" within the bubbling function.

The solution is here https://github.com/WordPress/gutenberg/pull/54080 

I've kept it on a separate PR/branch for us to compare but I'm personally more confident about #54080. it's less changes and keeps a very similar logic to trunk, only change is to switch the default event handler to an event handler on the body of the outer document.

youknowriad commented:
Closing this one in favor of #54080
